docs/cookbook/langgraph_vs_mirascope/quickstart.md

```diff
@@ -0,0 +1,390 @@
+            return None
+        else:
+            tool.query = input("(Assistant): Enter a new query: ")
+            return tool
+    else:
+        return tool
+
+def _step(self, question: str | None = None):
+    response = self._call(question)
+    tools_and_outputs = []
+    for chunk, tool in response:
+        if tool:
+            new_tool = self._interrupt_before(tool)
+            if new_tool:
+                tools_and_outputs.append((new_tool, new_tool.call()))
+            tools_and_outputs.append((tool, tool.call()))
+        else:
+            print(chunk.content, end="", flush=True)
+    if response.user_message_param:
+        self.history.append(response.user_message_param)
+    self.history.append(response.message_param)
+    if tools_and_outputs:
+        self.history += response.tool_message_params(tools_and_outputs)
+        return self._step()
+    return response.content
+
+```
+
+Now, before the LLM calls the tool, it will prompt the user requesting permission:
+
+```python
+Chatbot.run()
+(User): Can you tell me about the Mirascope python library?
+(Assistant): Do you want to use the tool? (y/n): y
+```
+
+### Human-as-a-tool
+
+We can use prompt engineering techniques to help the LLM make a decision on whether it needs human intervention or not. Let's add a `RequestAssistance` tool, and update our prompt so the LLM knows how to use our new tool.
+
+```python
+class RequestAssistance(openai.OpenAITool):
+    """A tool that requests assistance from a human expert."""
+
+    query: str = Field(..., description="The users question.")
+
+    def call(self) -> str:
+        """Prompts a human to enter a response."""
+        print(f"The AI has requested assistance. Here is the question: {self.query}")
+        response = input("(Human): ")
+        return f"Human response: {response}"
+
+class Chatbot(BaseModel):
+    ...
+
+    @openai.call(model="gpt-4o-mini", stream=True, tools=[WebSearch, RequestAssistance])
+    @prompt_template(
+        """
+        SYSTEM:
+        You are an expert web searcher. 
+        Your task is to answer the user's question using the provided tools.
+        You have access to the following tools:
+            - `WebSearch`: Search the web for information.
+            - `RequestAssistance`: Request assistance from a human expert if you do not
+                know how to answer the question.
+
+        Once you have gathered all of the information you need, generate a writeup that
+        strikes the right balance between brevity and completeness. The goal is to
+        provide as much information to the writer as possible without overwhelming them.
+
+        MESSAGES:
+        {self.history}
+
+        USER:
+        {question}
+        """
+    )
+    def _call(self, question: str | None = None): ...
+
+    ...
+
+Chatbot.run()
+# (User): What is my name?
+# (Assistant): The AI has requested assistance. Here is the question: What is my name?
+# (Human): The users name is Will.
```

Comment:
I wonder if we should either indent `(Human)` or re-print `(Assistant)` after the human's input? thinking there's something we can do here to make the print more readable, wdyt?

Reply:
I will indent human since technically thats hidden from the end user and is just a visual explanation of whats happening